### PR TITLE
drivers: clock_control: Add support for nRF54H20 clock controllers

### PR DESCRIPTION
**Pull request by anangl**
Add custom clock_control API for nRF platforms that allows requesting clocks with specified minimal required attributes (accuracy, precision, and frequency). Provide an implementation of this API for FLL16M, HFXO, HSFLL, and LFCLK controllers in the nRF54H20 SoC.

### CODE REVIEW

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+ * Definition of `struct clock_config_generic`.
+ * Used to access `clock_config_*` structures in a common way.
+ */
+STRUCT_CLOCK_CONFIG(generic, ONOFF_CNT_MAX);
+
+static void update_config(struct clock_config_generic *cfg)
+{
+	atomic_val_t prev_flags = atomic_or(&cfg->flags, FLAG_UPDATE_NEEDED);
+
+	if (prev_flags & (FLAG_UPDATE_NEEDED | FLAG_UPDATE_IN_PROGRESS)) {
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_common.c:29]: I'm confused here. It look like line 27 results in `prev_flags` equal to `FLAG_UPDATE_NEEDED` or `0`. So what is the purpose of checking `FLAG_UPDATE_IN_PROGRESS` here?

File: drivers/clock_control/clock_control_nrf2_fll16m.c [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+static void fll16m_work_handler(struct k_work *work)
+{
+	const struct device *hfxo = DEVICE_DT_GET(DT_NODELABEL(hfxo));
+	struct fll16m_dev_data *dev_data =
+		CONTAINER_OF(work, struct fll16m_dev_data, clk_cfg.work);
+	uint32_t active_options;
+
+	active_options = clock_config_update_begin(work);
+	if (active_options & BIT(2)) {
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_fll16m.c:56]: These magic bits in `active_options` are hard to follow in the review

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+#include "clock_control_nrf2_common.h"
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#include <hal/nrf_lrcconf.h>
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#define HFXO_START_TIME_US 900
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hfxo.c:16]: This value should be read from BICR. It's not constant. It depends on the installed xtal

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+struct dev_data_hfxo {
+	struct onoff_manager mgr;
+	onoff_notify_fn notify;
+	struct k_timer timer;
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hfxo.c:21]: Please note this timer is needed for FP1.0 revision. For FP1.1 it should not be necessary as a hardware event could be used instead.

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+};
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
+
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	struct hsfll_dev_data *dev_data = dev->data;
+
+	k_timer_stop(&dev_data->timer);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:48]: What if this function gets called after the timer fires? Is it ok to call `clock_config_update_end()` multiple times for one `_begin()`?

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+static const uint32_t max_frequency = MHZ(320);
+
+struct hsfll_dev_data {
+	STRUCT_CLOCK_CONFIG(hsfll, ARRAY_SIZE(oppoint_freq)) clk_cfg;
+	struct k_timer timer;
+};
+
+static void freq_setting_applied_cb(enum dvfs_frequency_setting new_setting)
+{
+	ARG_UNUSED(new_setting);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:43]: This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages. This function in this CPU should configure HSFLL to apply `new_setting`, even if it is higher than the last requested setting.

It can be fixed in another PR. But at least a comment like `TODO` would be good.

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+			       ARRAY_SIZE(dev_data->clk_cfg.onoff),
+			       hsfll_work_handler);
+	if (rc < 0) {
+		return rc;
+	}
+
+	k_timer_init(&dev_data->timer, hsfll_update_timeout_handler, NULL);
+
+	dvfs_service_handler_register_freq_setting_applied_callback(
+		freq_setting_applied_cb);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:218]: I expected to find requesting the initial frequency (the lowest one?) somewhere around here. Is it in some other place I missed?

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 8]
```diff
@@ -167,6 +179,28 @@ static int api_get_rate_hsfll(const struct device *dev,
 
 static int hsfll_init(const struct device *dev)
 {
+	const struct hsfll_dev_config *dev_config = dev->config;
+	nrf_hsfll_trim_t trim = {
+		.vsup   = sys_read32(dev_config->trim_vsup),
+		.coarse = sys_read32(dev_config->trim_coarse),
+		.fine   = sys_read32(dev_config->trim_fine),
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:186]: It looks like the `TCOEF` trim is missing from the list.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 6 above)
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:43]: And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

**Review by hubertmis — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_control_nrf2_common.c (hunk 1 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_common.c:29]: > I'm confused here. It look like line 27 results in `prev_flags` equal to `FLAG_UPDATE_NEEDED` or `0`.

No, `prev_flags` will have `FLAG_UPDATE_IN_PROGRESS` set if the update work is currently executed, and it will have `FLAG_UPDATE_NEEDED` if an update is already scheduled. In both these cases, the update work is not to be submitted (in the former case, it will be submitted at the end of `clock_config_update_end()`).

File: drivers/clock_control/clock_control_nrf2_fll16m.c (hunk 2 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_fll16m.c:56]: Yes, this will be reworked.

File: drivers/clock_control/clock_control_nrf2_common.c (hunk 1 above)
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_common.c:29]: I see. Too many reviews in a single day :)

**Comment by anangl** [comment]
Rebased.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 7 above)
Comment (Rafal-Nordic) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:240]: At least Secure Domain should set the lowest oppoint during initialisation. Without this, dvfs scaling on App core will not work.

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+#ifdef NRF_BICR
+	uint32_t bicr_startuptime_us = NRF_BICR->HFXO.STARTUPTIME;
+
+	if (bicr_startuptime_us != BICR_HFXO_STARTUPTIME_TIME_Unconfigured) {
+		dev_data->start_up_time = K_USEC(bicr_startuptime_us);
+	} else
+#endif
+	{
+		dev_data->start_up_time = K_USEC(1000);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hfxo.c:168]: BICR misconfiguration should be treated as a fault instead of setting a default value. This hardcoded default could be different than similar defaults in other libraries (e.g. radio protocols).

**Review by hubertmis — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 5 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:70]: `clock_config_update_end()` checks at its beginning if an update is in progress and returns if not. But indeed, it did not clear `FLAG_UPDATE_IN_PROGRESS` at that point, so another entry could cause issues in further processing. This has been fixed.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 6 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:65]: > And I'm missing the actual code modifying the frequency. Is it somewhere within this file?

It's handled in [ld_dvfs](https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs.c#L256).

Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:65]: > This part looks invalid. `freq_setting_applied_cb()` can be called unsolicited, when other CPUs are scaling their frequencies and voltages.

Yep. I'm reworking this.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 8 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:208]: Yes, it's missing. The nrfx HAL needs to be extended in this regard as currently the `nrf_hsfll_trim_t` structure does not contain the related field.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 5 above)
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:70]: Could you add API docs for the common module describing design principles like this? It would be easier to follow the programmers intention while reading the code.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 7 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:240]: Enabling `CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT` should do the job: https://github.com/zephyrproject-rtos/zephyr/blob/5d87cd654fbe2d430bcdaf39c666683abe65f290/modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c#L291-L294
I'll add this as default.

Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:240]: Now I see a race condition: the `ld_dvfs_something` module sets the initial frequency when ABB gets applied (max). In the meantime, this module is expected to set the initial frequency (probably min, or user-defined). These operations can happen in any order resulting in unknown initial frequency.

In the current design with `ld_dvfs_handler` I have a feeling that the initial frequency should be requested from clock_control to ld_dvfs_handler. The `ld_dvfs_handler` should apply the requested initial frequency immediately (if ABB has already been applied) or after ABB gets applied.

**Review by hubertmis — CHANGES_REQUESTED** [review]

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 7 above)
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:240]: Looks good. I've just spotted a potential race condition in ld_dvfs_handler.c:

The requested frequency is correctly set by the procedure you linked. However it might get incorrectly overwritten by `static void dvfs_service_handler_set_initial_hsfll_config(void)`. Is my observation correct?

File: modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c [hunk 10]
```diff
@@ -182,7 +213,13 @@ static void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)
 		break;
 	case NRFS_DVFS_EVT_OPPOINT_REQ_CONFIRMED:
 		/* Optional confirmation from sysctrl, wait for oppoint.*/
-		LOG_DBG("DVFS handler EVT_OPPOINT_REQ_CONFIRMED");
+		dvfs_service_handler_clear_state_bit(DVFS_SERV_HDL_FREQ_CHANGE_REQ_PENDING_BIT_POS);
+		LOG_DBG("DVFS handler EVT_OPPOINT_REQ_CONFIRMED %d", (uint32_t)p_evt->freq);
+		if (dvfs_service_handler_get_requested_oppoint() == p_evt->freq) {
```
[the remark this answers is not in the document]
Comment (lstnl) [reply modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:218]: This is case where requested frequency is same as previous one. If it is not then it will be handled later.

**Comment by anangl** [comment]
Rebased.

File: drivers/clock_control/clock_control_nrf2_lfclk.c [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+static void lfclk_work_handler(struct k_work *work)
+{
+	struct lfclk_dev_data *dev_data =
+		CONTAINER_OF(work, struct lfclk_dev_data, clk_cfg.work);
+	uint32_t active_options;
+	uint8_t option_idx = ARRAY_SIZE(clock_options) - 1;
+	nrfs_err_t err;
+
+	active_options = clock_config_update_begin(work);
+	while (option_idx) {
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_lfclk.c:86]: > can you use CLZ here to get highest bit that is set?

Isn't the optimizer smart enough? We shouldn't obscure the code with optimizations if not necessary

[the remark this answers is not in the document]
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_lfclk.c:86]: There is room for improvements in GCC (and clang apparently) ;)

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+{
+	struct dev_data_hfxo *dev_data =
+		CONTAINER_OF(timer, struct dev_data_hfxo, timer);
+
+	LOG_DBG("HFXOSTARTED: %u",
+		nrf_lrcconf_event_check(NRF_LRCCONF010,
+					NRF_LRCCONF_EVENT_HFXOSTARTED));
+
+	if (dev_data->notify) {
+		dev_data->notify(&dev_data->mgr, 0);
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/clock_control/clock_control_nrf2_hfxo.c:46]: i see in the comment that it may not be started.

File: modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c [hunk 13]
```diff
@@ -152,11 +165,29 @@ static void dvfs_service_handler_set_initial_hsfll_config(void)
...
 }
 
+/* Timer to add additional delay to finish downscale procedure when domain other than secure */
+#if !defined(NRF_SECURE)
+#define SCALING_FINISH_DELAY_TIMEOUT_US                                                            \
+	K_USEC(CONFIG_NRFS_LOCAL_DOMAIN_DOWNSCALE_FINISH_DELAY_TIMEOUT_US)
+
+static void dvfs_service_handler_scaling_finish_delay_timeout(struct k_timer *timer)
+{
+	if (timer) {
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c]: it was part of #73783 and no longer in that pr.

File: modules/hal_nordic/nrfs/dvfs/Kconfig [hunk 14]
```diff
@@ -26,6 +26,11 @@ config NRFS_LOCAL_DOMAIN_DOWNSCALE_SAFETY_TIMEOUT_US
 	default 1000000 if (NRFS_LOCAL_DOMAIN_DVFS_TEST || LOG)
 	default 1500
 
+config NRFS_LOCAL_DOMAIN_DOWNSCALE_FINISH_DELAY_TIMEOUT_US
+	int "Additional delay to let secdom finish dowscale procedure in us"
+	range 1 10000000
+	default 1000
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply modules/hal_nordic/nrfs/dvfs/Kconfig]: it was part of #73783 and no longer in that pr.

File: modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c [hunk 15]
```diff
@@ -152,11 +165,29 @@ static void dvfs_service_handler_set_initial_hsfll_config(void)
...
+		dvfs_service_handler_scaling_finish(
+			*(enum dvfs_frequency_setting *)timer->user_data);
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c]: it was part of #73783 and no longer in that pr.

File: modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c [hunk 16]
```diff
@@ -196,15 +233,20 @@ static void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)
 			dvfs_service_handler_scaling_background_job(p_evt->freq);
 			LOG_DBG("DVFS handler EVT_OPPOINT_SCALING_PREPARE handled");
 #if !defined(NRF_SECURE)
-			current_freq_setting = p_evt->freq;
+			/* Additional delay for downscale to finish on secdom side */
+			static enum dvfs_frequency_setting freq;
+
+			freq = p_evt->freq;
+			dvfs_service_scaling_finish_delay_timer.user_data = (void *)&freq;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c]: it was part of #73783 and no longer in that pr.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 5 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c]: Added.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 7 above)
Comment (anangl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:246]: From what I can tell, `dvfs_service_handler_task()` waits for `dvfs_service_sync_sem` before calling `dvfs_service_handler_change_freq_setting()`, so `dvfs_service_handler_set_initial_hsfll_config()` needs to complete first. Unless `NRFS_DVFS_EVT_INIT_DONE` is received twice (what would mean that something is really off), it should be fine.

Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:246]: Right. I didn't realize that the only invocations of these functions are protected from races with a semaphore. It's not the cleanest code (adding more calls to these functions can cause race conditions), but now it seems fine.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 8 above)
Comment (hubertmis) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:214]: Do you plan to fix this in this PR?

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+	struct clock_config_generic *cfg =
+		CONTAINER_OF(work, struct clock_config_generic, work);
+	uint32_t active_options;
+
+	(void)atomic_or(&cfg->flags, FLAG_UPDATE_IN_PROGRESS);
+	cfg->flags_snapshot = atomic_and(&cfg->flags, ~FLAG_UPDATE_NEEDED);
+
+	active_options = cfg->flags_snapshot & BIT_MASK(ONOFF_CNT_MAX);
+
+	return active_options ? (uint8_t)(31 - __builtin_clz(active_options))
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_common.c:112]: I would move `(uint8_t)(31 - __builtin_clz(active_options))` to a `static inline` function with a name indicating what it is intended to calculate. Right now I'm guessing what was your intention :)

**Review by hubertmis — COMMENTED** [review]
The missing TCOEF in HSFLL trimming is the last thing I would like to agree on before I approve. It's fine for me to implement it out of this PR, but we need at least to track it to make sure we don't forget

File: drivers/clock_control/clock_control_nrf2_fll16m.c [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+};
+
+struct fll16m_dev_config {
+	uint32_t fixed_frequency;
+};
+
+static void activate_fll16m_mode(struct fll16m_dev_data *dev_data, uint8_t mode)
+{
+	/* TODO: change to nrf_lrcconf_* function when such is available. */
+	NRF_LRCCONF010->CLKCTRL[0].SRC = mode;
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_fll16m.c:63]: While debugging a clock issue with @KAGA164 I realized we've forgotten to manage LRCCONF's power. In some scenarios, it results in losing clock mode requests.

Setting `mode` to any other value than the hardware default (open-loop) should be preceded by `NRF_LRCCONF010->POWERON.MAIN = 1`. Setting `mode` to the hardware default should be followed by resetting `POWERON.MAIN`.

Please note, that other components could require setting POWERON.ACT or POWER.MAIN independently. One example is HFXO clock_control instance. So accessing LRCCONF010->POWERON.MAIN requires a multi-client helper (another on-off instance?)

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+static void onoff_start_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	struct dev_data_hfxo *dev_data =
+		CONTAINER_OF(mgr, struct dev_data_hfxo, mgr);
+	const struct device *dev = DEVICE_DT_INST_GET(0);
+	const struct dev_config_hfxo *dev_config = dev->config;
+
+	dev_data->notify = notify;
+
+	nrf_lrcconf_event_clear(NRF_LRCCONF010, NRF_LRCCONF_EVENT_HFXOSTARTED);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hfxo.c:59]: Like described in FLL16M, this must be preceded by setting `NRF_LRCCONF010->POWERON.MAIN = 1`.

File: drivers/clock_control/clock_control_nrf2_hfxo.c [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+	/* Due to a hardware issue, the HFXOSTARTED event is currently
+	 * unreliable. Hence the timer is used to simply wait the expected
+	 * start-up time. To be removed once the hardware is fixed.
+	 */
+	k_timer_start(&dev_data->timer, dev_config->start_up_time, K_NO_WAIT);
+}
+
+static void onoff_stop_hfxo(struct onoff_manager *mgr, onoff_notify_fn notify)
+{
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_STOPREQHFXO);
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hfxo.c:71]: And this must be followed with `NRF_LRCCONF010->POWERON.MAIN = 0` if no other user requires keeping the main power domain on. The "If no other user..." part requires users counting, like another on-off instance?

**Review by hubertmis — CHANGES_REQUESTED** [review]
I've just realized there is overlooked bug in this implementation.

**Comment by bjarki-andreasen** [comment]
This PR will be continued from the new PR #77365

File: drivers/clock_control/clock_control_nrf2_common.c (hunk 17 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_common.c:112]: Done in new PR

File: drivers/clock_control/clock_control_nrf2_fll16m.c (hunk 18 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_fll16m.c:63]: Added in new PR

File: drivers/clock_control/clock_control_nrf2_hfxo.c (hunk 19 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_hfxo.c:59]: Added in new PR

File: drivers/clock_control/clock_control_nrf2_hfxo.c (hunk 20 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_hfxo.c:71]: Added in new PR

File: dts/common/nordic/nrf54h20.dtsi [hunk 21]
```diff
@@ -723,6 +742,7 @@
 				reg = <0x99c000 0x1000>;
 				status = "disabled";
 				cc-num = <16>;
+				clocks = <&lfclk>;
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/common/nordic/nrf54h20.dtsi:826]: Added in new PR

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 7 above)
Comment (lstnl) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:246]: CONFIG_NRFS_LOCAL_DOMAIN_DVFS_SCALE_DOWN_AFTER_INIT this config was introduced to enable option to change DVFS settings from application core without other existing solution to do that. I don't know if this shouldn't be explicitly called by secdom for example when all initialization is done?
